Add `ray cluster-dump` CLI command to fetch logs

krfricke opened the pull request:
…a cluster

<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Collect logs from nodes for debugging.

Questions:
- Naming of the command
- Do we have/need unit tests for CLI commands?

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

krfricke commented:
Currently this will add `dataclasses` as a setup.py dependency (dataclasses was merged upstream in python >=3.7).

richardliaw commented:
Can we avoid the dataclasses dependency and make this call available for only 3.7 and above? 

The dataclasses dependency has often caused issues (and we want to reduce our dependency footprint)

krfricke commented:
Ok, done. I can also refactor the PR to not use dataclasses

richardliaw commented on python/ray/scripts/scripts.py:
```diff
@@ -19,6 +21,15 @@
     attach_cluster, exec_cluster, create_or_update_cluster, monitor_cluster,
     rsync, teardown_cluster, get_head_node_ip, kill_node, get_worker_node_ips,
     debug_status, RUN_ENV_TYPES)
+try:
+    from ray.autoscaler._private.get_logs import Archive, \
+        GetParameters, Node, _info_from_params, \
+        create_archive_for_local_and_remote_nodes, \
+        create_archive_for_remote_nodes, get_all_local_data
+    get_logs_available = True
+except ImportError:
+    get_logs_available = False
```
Hmmm, maybe it is cleaner just to refactor to not have `dataclasses`

richardliaw commented on python/ray/scripts/scripts.py:
```diff
@@ -19,6 +21,15 @@
     rsync, teardown_cluster, get_head_node_ip, kill_node, get_worker_node_ips,
     debug_status, RUN_ENV_TYPES)
+try:
+    from ray.autoscaler._private.get_logs import Archive, \
+        GetParameters, Node, _info_from_params, \
+        create_archive_for_local_and_remote_nodes, \
+        create_archive_for_remote_nodes, get_all_local_data
+    get_logs_available = True
```
I just noticed that this introduces a lot of imports into this file. Can you push the logic for `get-logs` and `get-local-logs` into a function in `autoscaler._private.commands`? I think we have tended towards pushing most of the logic in that file

rkooo567 commented:
When I proposed a Ray log API, the naming proposal was `ray log --archive-all`. Why don't we just go with this now?

wuisawesome commented on python/ray/scripts/scripts.py:
```diff
@@ -1414,6 +1427,235 @@ def status(address, redis_password):
+    is_flag=True,
+    default=True,
+    help="Increase process information verbosity")
+def get_logs(cluster_config_file: Optional[str] = None,
```
Can we call it `core-dump`?

rkooo567 replied:
Hmm I am seeing our core dump to be more internal information (e.g., ray memory output, some debug strings, and etc.) rather than dump of logs (though I don't object your idea to call it core dump and include logs eventually).

wuisawesome replied:
isn't the point/use-case of this supposed to be for getting internal information? (I think we should add everything you mentioned to this command :p )

wuisawesome commented:
From a privacy/transparency perspective: I wonder if we should document exactly what we're including in the archive...

krfricke commented:
It's probably a good idea to state that in the CLI output.

I'll refactor the dataclasses in the next days!

krfricke commented:
I renamed the command to `cluster-dump`, as I think this fits quite good.

I also removed the dependency on dataclasses, refactored the command into `commands.py` and added a warning about which logs are included in the archive.

krfricke commented:
Example output:

```
 ray cluster-dump ../nightly.yaml
2021-02-22 11:21:57,296	WARNING commands.py:1198 -- You are about to create a cluster dump. This will collect data from cluster nodes.

The dump will contain this information:

  - The logfiles of you Ray session
    This usually includes Python outputs (stdout/stderr)
  - Your installed Python packages (`pip freeze`)
  - Information on your running Ray processes
    This includes command line arguments

If you are concerned about leaking private information, extract the archive and inspect its contents before sharing it with anyone.
2021-02-22 11:21:57,296	INFO cluster_dump.py:514 -- Retrieving cluster information from ray cluster file: ../nightly.yaml
2021-02-22 11:21:57,474	INFO commands.py:294 -- Checking AWS environment settings
2021-02-22 11:22:12,298	INFO cluster_dump.py:368 -- Collecting data from remote node: 18.xxx.xxx.xxx
2021-02-22 11:22:12,299	INFO cluster_dump.py:368 -- Collecting data from remote node: 34.xxx.xxx.xxx
2021-02-22 11:22:12,299	INFO cluster_dump.py:368 -- Collecting data from remote node: 35.xxx.xxx.xxx
Created archive: ./ray-get-logs/tmp/ray-gather-logs-nightly_2021-02-22_11-22-12.tar.gz
```

richardliaw commented on python/ray/autoscaler/_private/cluster_dump.py:
```diff
@@ -0,0 +1,599 @@
+        Open archive object.
+    """
+    if not processes:
+        # local import to avoid circular dependencies
+        from ray.scripts.scripts import RAY_PROCESSES
+        processes = RAY_PROCESSES
```
Could we put this in ray.autoscaler.constants?

richardliaw commented on python/ray/autoscaler/_private/cluster_dump.py:
```diff
@@ -0,0 +1,599 @@
+                processes=parameters.processes_list,
+                verbose=parameters.processes_verbose)
+        except LocalCommandFailed as exc:
+            logger.error(exc)
```
in the future, it'd be great to switch to the autoscaler CLI

richardliaw reviewed (approved):
Added a nit -- otherwise looks great to me!

rkooo567 commented:
Hey @krfricke Can we also collect debug_state.txt as well? I think this could be pretty useful.

richardliaw commented:
@rkooo567 let me open a github issue for your request and just merge this in first!
